girder_worker/docker/transforms/__init__.py

```diff
@@ -0,0 +1,278 @@
+        """
+        super(TemporaryVolume, self).__init__(None, None)
+        self.host_dir = host_dir
+        self._instance = None
+        self._transformed = False
+
+    def transform(self, **kwargs):
+        if not self._transformed:
+            self._transformed = True
+            self._make_paths(self.host_dir)
+
+        return super(TemporaryVolume, self).transform(**kwargs)
+
+
+class _DefaultTemporaryVolume(TemporaryVolume):
+    """
+    Place holder who delegates implementation to instance provide by transform(...) method
+    An instance of the class is returned each time `TemporaryVolume.default` is accessed.
+    When the docker_run task is executed the transform(...) method is call with an instance
+    containing information about the actual default temporary volume associated with the
+    task. The place holder then delgates all functionality to this instance.
+    """
+    def transform(self, _default_temp_volume=None, **kwargs):
+        self._instance = _default_temp_volume
+        self._transformed = True
+
+        return self._instance.transform(**kwargs)
+
+    @property
+    def container_path(self):
+        return self._instance.container_path
+
+    @property
+    def host_path(self):
+        return self._instance.host_path
+
+
+class NamedPipeBase(Transform):
+    def __init__(self, name, container_path=None, host_path=None, volume=TemporaryVolume.default):
+        super(NamedPipeBase, self).__init__()
+        self._container_path = None
+        self._host_path = None
+        self._volume = None
+
+        if container_path is not None and host_path is not None:
+            self._container_path = container_path
+            self._host_path = host_path
+        else:
+            self._volume = volume
+
+        self.name = name
+
+    def transform(self, **kwargs):
+        if self._volume is not None:
+            self._volume.transform(**kwargs)
+
+    @property
+    def container_path(self):
+        if self._container_path is not None:
+            return os.path.join(self._container_path, self.name)
+        else:
+            return os.path.join(self._volume.container_path, self.name)
+
+    @property
+    def host_path(self):
+        if self._host_path is not None:
+            return os.path.join(self._host_path, self.name)
+        else:
+            return os.path.join(self._volume.host_path, self.name)
+
+    def cleanup(self, **kwargs):
+        os.remove(self.host_path)
+
+
+class NamedInputPipe(NamedPipeBase):
+    """
+    A named pipe that read from within a docker container.
+    i.e. To stream data out of a container.
+    """
+    def __init__(self, name,  container_path=None, host_path=None, volume=TemporaryVolume.default):
+        super(NamedInputPipe, self).__init__(name, container_path, host_path, volume)
+
+    def transform(self, **kwargs):
+        from girder_worker.docker.io import (
+            NamedPipe,
+            NamedPipeWriter
+        )
```

Comment:
I think we can lift out these imports too

Reply:
Same as above